### PR TITLE
feat(mcp): add results pagination cache and MRTR

### PR DESCRIPTION
**Pull request by jolestar**
## What

Add MCP result normalization, complete list pagination, scoped schema caching, and Multi Round-Trip Request (MRTR) continuation support for the modern protocol while preserving legacy behavior.

## Why

This is PR 3 of the ordered MCP 2026-07-28 compatibility series. Modern MCP responses introduce typed result variants and continuation flows, while catalog discovery must handle pagination and cache isolation across protocol/auth scopes.

## How

- Normalize modern and legacy MCP results behind a stable `resultType` output contract.
- Follow cursors across tools, prompts, resources, and resource-template catalog requests.
- Scope schema cache metadata by endpoint, protocol era, and authorization identity.
- Invalidate matching memory and disk schema caches on list-change notifications.
- Add explicit MRTR pending-result output and continuation controls across stdio, HTTP, and daemon reuse paths.

## Changes

- [x] Add MRTR execution and continuation controls.
- [x] Paginate MCP list catalogs.
- [x] Scope schema cache metadata and notification invalidation.
- [x] Extend regression and integration coverage for modern/legacy behavior.

### Unit Tests
- [x] All existing tests pass
- [x] New tests added
- [x] `cargo test` passes

### Manual Testing
- [x] Tested transport continuation paths
- [x] Tested pagination and cache edge cases
- [x] Tested error and stale-cache fallback conditions

### Test Results

```bash
cargo +1.91.1 test --locked -- --test-threads=1
cargo +1.91.1 clippy --locked --lib --bins -- -D warnings
```

## Checklist

- [x] Code formatted with `cargo fmt`
- [x] No clippy warnings for library and binary targets
- [x] All tests pass
- [ ] Documentation updated (planned for PR 5)
- [x] Commit messages follow convention
- [x] PR title follows convention

## Breaking Changes

None. Legacy MCP behavior remains supported; modern result details are exposed through additive output fields and explicit continuation controls.

## Additional Notes

This PR depends on the protocol foundation and modern Streamable HTTP support merged in PRs #428 and #429. Subscriptions/daemon listening and OAuth/conformance documentation remain scoped to PRs 4 and 5.

### CODE REVIEW

File: src/daemon.rs [hunk 1]
```diff
@@ -8530,6 +8555,21 @@ async fn resolve_adapter_with_schema_cache(
...
 }
 
+fn lookup_schema_cache(
+    cache: &dyn cache::Cache,
+    url: &str,
+    auth_profile: Option<&Profile>,
+    policy: cache::CacheReadPolicy,
+) -> Result<Option<(cache::CacheHit, String)>> {
+    let mcp_cache_key = adapters::mcp::McpAdapter::schema_cache_key_for(url, auth_profile);
+    for cache_key in [url.to_string(), mcp_cache_key] {
```
[the remark this answers is not in the document]
Comment (jolestar) [reply src/daemon.rs]: 已在提交 ef5cee3 中修复：MCP schema cache 现在优先读取 auth-scoped key，仅在未命中时回退 legacy plain-URL key，并补充了对应回归测试。

File: src/adapters/mcp/client.rs [hunk 2]
```diff
@@ -515,15 +523,145 @@ impl McpStdioClient {
...
+        capabilities: Option<&JsonValue>,
+        timeout: Duration,
     ) -> Result<JsonValue> {
         let params = match self.protocol_context.era {
-            McpProtocolEra::Modern => Some(self.protocol_context.modern_request_params(params)?),
+            McpProtocolEra::Modern => Some(
+                self.protocol_context
+                    .modern_request_params_with_capabilities(params, capabilities)?,
+            ),
             McpProtocolEra::Legacy => params,
```
[the remark this answers is not in the document]
Comment (jolestar) [reply src/adapters/mcp/client.rs:542]: 已在提交 3a52971 中处理：抽取共享 CatalogPaginationState，将四处循环统一为同一分页状态机，并保留各 transport 的请求发送差异；相关 MCP 单元测试与 daemon 回归测试均已通过。

**Comment by jolestar** [comment]
@holonbot review

File: src/adapters/mcp/mod.rs [hunk 3]
```diff
@@ -79,6 +115,53 @@ impl McpAdapter {
...
+            }
+        } else {
+            hasher.update(b"anonymous");
+        }
+        format!(
+            "{}#uxc-mcp-schema=dual-era-2026&auth={:x}",
+            url,
+            hasher.finalize()
+        )
+    }
```
[the remark this answers is not in the document]
Comment (jolestar) [reply src/adapters/mcp/mod.rs:136]: 已在提交 746e846 中修复：auth profile 先转换为 JSON Value，再对 object key 递归排序后计算 canonical hash；新增字段插入顺序不同但逻辑相同的回归测试。

File: src/daemon.rs [hunk 4]
```diff
@@ -1019,7 +1021,7 @@ impl McpStdioSession {
             if notification.method == "notifications/tools/list_changed" {
                 self.tools_dirty = true;
                 if let Some(cache) = cache {
-                    let _ = cache.invalidate(endpoint);
+                    let _ = cache.invalidate(&self.schema_cache_key);
                 }
```
[the remark this answers is not in the document]
Comment (jolestar) [reply src/daemon.rs:1025]: 已在提交 746e846 中修复：tools/list_changed 同时失效 auth-scoped key 与兼容回退使用的 legacy endpoint key，并新增双 key invalidation 回归测试。

File: src/daemon.rs [hunk 5]
```diff
@@ -8777,13 +8821,33 @@ async fn prepare_runtime_execute_args(
...
+                .into_iter()
+                .chain(
+                    continuation
+                        .map(|value| (adapters::mcp::MCP_CONTINUATION_ARG.to_string(), value)),
+                ),
+        ))?;
+        Some(options)
+    } else {
+        None
+    };
```
[the remark this answers is not in the document]
Comment (jolestar) [reply src/daemon.rs]: 已在提交 746e846 中修复：daemon 在进入任意 adapter 的参数准备前统一剥离 MCP control fields；仅 MCP adapter 会校验并重新注入 execution options，并新增 non-MCP 回归测试。

File: src/cache/storage.rs [hunk 6]
```diff
@@ -474,6 +477,8 @@ impl Cache for SchemaCache {
 mod tests {
     use super::*;
     use std::fs;
+    use std::thread::sleep;
+    use std::time::Duration;
     use tempfile::TempDir;
```
[the remark this answers is not in the document]
Comment (jolestar) [reply src/cache/storage.rs:482]: 已核查：当前 src/cache/storage.rs 的 TTL 回归测试仍实际调用 sleep(Duration::from_secs(2))，因此这两个 import 不是 unused；最新提交的 cargo clippy -- -D warnings 已通过。本 PR 暂不改动该独立的非阻塞测试优化。

**Comment by jolestar** [comment]
@holonbot review

**Comment by jolestar** [comment]
@copilot review

**Comment by jolestar** [comment]
@holonbot review

**Comment by jolestar** [comment]
@copilot review
